lang/src/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransform.kt

```diff
@@ -530,120 +529,74 @@ internal class StaticTypeInferenceVisitorTransform(
                     )
                 )
             )
         }
 
-        private fun computeReturnTypeForAggFunc(aggFunc: SymbolPrimitive, argType: StaticType, sourceLocation: SourceLocationMeta): StaticType {
-            // nested type is not supported by agg functions.
-            // i.e. sum(1,2,[3,4]) is not supported
-            val isNullOrMissingOrNumeric: (StaticType) -> Boolean = {
-                it.isSubTypeOf(StaticType.unionOf(StaticType.MISSING, StaticType.NULL, StaticType.NUMERIC))
+        private fun computeReturnTypeForAggFunc(funcName: String, elementType: StaticType, sourceLocation: SourceLocationMeta): StaticType {
+            if (funcName == "count") {
+                return StaticType.INT
             }
 
-            return when (aggFunc.text) {
-                // current implementation of count will always return a long
-                "count" -> StaticType.INT8
-                // max/min supports all type and the result depends on comparison.
-                // aggregate function will not return missing as a potential Type
-                // in case that argType contains only missing,
-                "max", "min" -> {
-                    val possibleReturnTypes = argType.allTypes.filter {
-                        it !is MissingType
-                    }
-                    if (possibleReturnTypes.isEmpty())
-                        StaticType.NULL
-                    else
-                        StaticType.unionOf(possibleReturnTypes.toSet()).flatten()
+            val elementTypes = elementType.allTypes
+
+            fun List<StaticType>.convertMissingToNull() = toMutableSet().apply {
+                if (contains(StaticType.MISSING)) {
+                    remove(StaticType.MISSING)
+                    add(StaticType.NULL)
                 }
-                // current implementation of avg always return a decimal or null.
-                "avg" -> {
-                    when {
-                        !isNullOrMissingOrNumeric(argType) -> {
-                            handleInvalidInputTypeForAggFun(sourceLocation, aggFunc, argType, StaticType.unionOf(StaticType.MISSING, StaticType.NULL, StaticType.NUMERIC))
-                            StaticType.unionOf(StaticType.NULL, StaticType.DECIMAL)
-                        }
-                        // missing, null, or missing and null
-                        argType.isSubTypeOf(StaticType.unionOf(StaticType.MISSING, StaticType.NULL)) -> StaticType.NULL
-                        else -> StaticType.DECIMAL
+            }
+
+            fun StaticType.isUnknownOrNumeric() = isUnknown() || isNumeric()
+
+            return when (funcName) {
+                // In case that any element is MISSING or there is no element, we should return NULL
+                "max", "min" -> StaticType.unionOf(elementTypes.convertMissingToNull())
+                "sum" -> when {
+                    elementTypes.all { !it.isUnknownOrNumeric() } -> {
```

Comment:
nit: boolean expression could be less confusing with the following:

```suggestion
                    elementTypes.none { it.isUnknownOrNumeric() } -> {
```